entity-server-api/src/main/java/org/terracotta/entity/PassiveServerEntity.java

```diff
@@ -36,9 +36,15 @@
    *  under a given key, the messages it creates and sends for synchronization will be executed in that key, on the passive,
    *  via this invoke() entry-point.</p>
    *
+   * @param clientDescriptor source instance from which the invocation originates.
+   * @param currentOrderedId current id this invoke is part of
+   * @param eldestOrderedId earliest active id for this client
    * @param message The message from a client or upstream active server
    */
-  void invoke(M message) throws EntityUserException;
+  void invoke(ClientDescriptor clientDescriptor,
```

Comment:
The additional information allows you to associate messages (invokes) with a specific client/transaction. For an entity, this can be used to handle the reconnect case. On reconnect (failover from active to passive), the client will resend all it's outstanding messages. It is possible for the client to resend a message that the newly promoted active has already applied. This information will allow for entities to track ignore already applied messages, if that is important. I believe Albin & Abhilash are building the common tracking functionality on top of this. 

Also, this is only of concern if your entity has non-idempotent operations. If you are fine with reapplying some messages twice on failover, then you can completely ignore this.